cmd/maya-apiserver/cstor-operator/cspc/controller.go

```diff
@@ -0,0 +1,219 @@
+	}
+}
+
+// withKubeClient fills kube client to controller object.
+func (cb *ControllerBuilder) withKubeClient(ks kubernetes.Interface) *ControllerBuilder {
+	cb.Controller.kubeclientset = ks
+	return cb
+}
+
+// withOpenEBSClient fills openebs client to controller object.
+func (cb *ControllerBuilder) withOpenEBSClient(cs clientset.Interface) *ControllerBuilder {
+	cb.Controller.clientset = cs
+	return cb
+}
+
+// withNDMClient fills ndm client to controller object.
+func (cb *ControllerBuilder) withNDMClient(ndmcs ndmclientset.Interface) *ControllerBuilder {
+	cb.Controller.ndmclientset = ndmcs
+	return cb
+}
+
+// withSpcLister fills cspc lister to controller object.
+func (cb *ControllerBuilder) withCSPCLister(sl informers.SharedInformerFactory) *ControllerBuilder {
+	cspcInformer := sl.Openebs().V1alpha1().CStorPoolClusters()
+	cb.Controller.cspcLister = cspcInformer.Lister()
+	return cb
+}
+
+// withcspcSynced adds object sync information in cache to controller object.
+func (cb *ControllerBuilder) withCSPCSynced(sl informers.SharedInformerFactory) *ControllerBuilder {
+	cspcInformer := sl.Openebs().V1alpha1().CStorPoolClusters()
+	cb.Controller.cspcSynced = cspcInformer.Informer().HasSynced
+	return cb
+}
+
+// withWorkqueue adds workqueue to controller object.
+func (cb *ControllerBuilder) withWorkqueueRateLimiting() *ControllerBuilder {
+	cb.Controller.workqueue = workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "CSPC")
+	return cb
+}
+
+// withRecorder adds recorder to controller object.
+func (cb *ControllerBuilder) withRecorder(ks kubernetes.Interface) *ControllerBuilder {
+	glog.V(4).Info("Creating event broadcaster")
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&typedcorev1.EventSinkImpl{Interface: ks.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, corev1.EventSource{Component: controllerAgentName})
+	cb.Controller.recorder = recorder
+	return cb
+}
+
+// withEventHandler adds event handlers controller object.
+func (cb *ControllerBuilder) withEventHandler(cspcInformerFactory informers.SharedInformerFactory) *ControllerBuilder {
+	cspcInformer := cspcInformerFactory.Openebs().V1alpha1().CStorPoolClusters()
+	// Set up an event handler for when CSPC resources change
+	cspcInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    cb.Controller.addSpc,
+		UpdateFunc: cb.Controller.updateSpc,
+		// This will enter the sync loop and no-op, because the cspc has been deleted from the store.
+		DeleteFunc: cb.Controller.deleteSpc,
+	})
+	return cb
+}
+
+// Build returns a controller instance.
+func (cb *ControllerBuilder) Build() (*Controller, error) {
+	err := openebsScheme.AddToScheme(scheme.Scheme)
+	if err != nil {
+		return nil, err
+	}
+	return cb.Controller, nil
+}
+
+// addSpc is the add event handler for cspc
+func (c *Controller) addSpc(obj interface{}) {
```

Comment:
spc name need to be changed

Reply:
addSpc/updateSpc/enqueueSpc/oldSpc/newSpc etc

Reply:
will fix it